_source/_posts/2022-07-06-spa-web-security.md

```diff
@@ -0,0 +1,137 @@
+
+![OWASP Top 10 for 2021 shows the following vulnerabilities in order: Broken Access Control, Cryptographic Failures, Injection, Insecure Design, Security Misconfiguration, Vulnerable and Outdated Components, Identification and Authentication Failures, Software and Data Integrity Failures, Security Logging and Monitoring Failures, and Server-Side Request Forgery](https://owasp.org/www-project-top-ten/assets/images/mapping.png)
+
+As this list is a list of web vulnerabilities, the entire list is pertinent to our work as web developers. However, a few critical vulnerabilities call for extra attention, and some vulnerabilities are less susceptible when the front-end is a SPA with a JSON-based backend API.
+
+We will cover some of these vulnerabilities specifically, but I'll also list general things to keep in mind that feed into the vulnerabilities listed here.
+
+## Leverage your browser and security headers
+First, we need to set the stage for part of our vulnerability mitigation strategy by bringing up security headers. Modern browsers (buh-bye Internet Explorer!!) have a lot of security mechanisms already built into them. We have to leverage the built-in security mechanisms and can enhance them with additional security headers. Security headers are HTTP response headers that provide instructions on how the browser should behave and are defined on your web server.
+
+There are quite a few security headers that we can apply. Something to note, SPAs are unique, so not all security headers apply in the same way you would for a traditional server-rendered web application. I'll call out the security headers that you'll want to use when we discuss how to mitigate a specific vulnerability. 
+
+### Dive deeper into security headers
+Here are some great resources to learn more about security headers. Figuring out how to configure the security headers can be esoteric. I like these posts that are short and actionable steps to configure security headers. 
+* [An Overview of Best Practices for Security Headers](/blog/2021/10/18/security-headers-best-practices)
+* [OWASP's HTTP Security Response Headers Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/HTTP_Headers_Cheat_Sheet.html)
+* [Web.dev's Security headers quick reference](https://web.dev/security-headers/)
+
+## Securely transport data using TLS/SSL
+Using HTTPS should be the standard without needing to say anything, but there are legacy systems where applications don't use HTTPS. It's essential to use the _secure_ hypertext transfer protocol; after all, the HTTPS acronym is appropriately derived from **H**ypertext **T**ransfer **P**rotocol **S**ecure. And it's essential to be secure across your _entire_ stack, not just when serving the front-end site (don't ask me how I know this can be a thing). 
```

Comment:
```suggestion
Using HTTPS should be the undisputed standard, but there are legacy systems where applications don't use HTTPS. It's essential to use the _secure_ hypertext transfer protocol; after all, the HTTPS acronym is appropriately derived from **H**ypertext **T**ransfer **P**rotocol **S**ecure. And it's essential to be secure across your _entire_ stack, not just when serving the front-end site (don't ask me how I know this can be a thing). 
```
terser